### ISSUE TITLE
Use assertEqual etc unittest methods

### ISSUE THREAD

**Issue by peterjc** [biopython/biopython#2835]
See #2700 for background, cross reference https://github.com/biopython/biopython/pull/2833#discussion_r406737022 e.g.

```
$ grep assertTrue *.py | grep "=="
common_BioSQL.py:        self.assertTrue(self.db.load(seqiter) == 1)
test_ClustalOmega_tool.py:        self.assertTrue(error.strip() == "" or
...
```

and:

```
$ grep assertFalse *.py | grep "!="
test_Restriction.py:        self.assertFalse(Acc65I != Asp718I)
```

Those should use ``assertEqual``, while:

```
$ grep assertTrue *.py | grep "!="
test_SearchIO_model.py:        self.assertTrue(hit11.id != "alt1")
test_SearchIO_model.py:        self.assertTrue(hit11.id != "alt11")
test_SearchIO_model.py:        self.assertTrue(hit11.id != "alt1")
```

and:

```
$ grep assertFalse *.py | grep "=="
test_PDB.py:        self.assertFalse(struct[0] == struct[1])
test_PDB.py:        self.assertFalse(struct[0] == [])  # __eq__ diff. types
test_PDB.py:        self.assertFalse(struct == model)
test_PDB.py:        self.assertFalse(res1 == res2)
test_PDB_StructureAlignment.py:        self.assertFalse(al.map12 == al.map21)
test_Restriction.py:        self.assertFalse(FormattedSeq(Seq("GATC")) ==
test_Restriction.py:        self.assertFalse(FormattedSeq(Seq("TAGC")) == Seq("TAGC"))
test_Restriction.py:        self.assertFalse(Acc65I == Asp718I)
```

Those shoud use ``self.assertNotEqual``

Not this cleanup work is not specific to dropping Python 3:

https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual
https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertNotEqual

However, some of the older usage probably did require at least Python 2.7:

https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertEqual

**Comment by peterjc** [comment]
See also https://github.com/biopython/biopython/issues/2700#issuecomment-594032554 where the ``A500`` check from ``flake8-assertive`` should catch more than my quick grep searches.

**Comment by trkohler** [comment]
Hi!
Can I work on this? I'd like to help! 
It might be my first open-source issue so if it requires any special knowledge, please let me know

**Comment by peterjc** [comment]
@trkohler Welcome! You should be fine with basic Python knowledge (very few of the Biopython objects do anything strange with self-defined equality rules). I suggest you pick a related batch of files, say which ones here (to reduce the chance of someone else working on the same thing), and start working. Thank you.

**Comment by peterjc** [comment]
To clarify, probably best don't touch ``test_Restriction.py`` which was done in #2833.

**Comment by trkohler** [comment]
may I consider this one as done? Or did I miss something again? I'm sorry for my inattention, I was too excited by the fact I can do something for such a project.

**Comment by peterjc** [comment]
Using the ``flake8-assertive`` plugin, code A500 covers more than just equal and not equal, but:

```
$ flake8 --select A500 .
...
./test_Pathway.py:80:9: A500 prefer assertEqual() for '==' comparisons
./test_Pathway.py:82:9: A500 prefer assertEqual() for '==' comparisons
./test_Pathway.py:160:9: A500 prefer assertEqual() for '==' comparisons
...
./test_PDB.py:682:9: A500 prefer assertEqual() for '==' comparisons
./common_BioSQL.py:1212:9: A500 prefer assertEqual() for '==' comparisons
...
./test_Restriction.py:233:9: A500 prefer assertNotEqual() for '==' comparisons
./test_Restriction.py:235:9: A500 prefer assertEqual() for '!=' comparisons
./test_Restriction.py:242:9: A500 prefer assertGreaterEqual() for '>=' comparisons
./test_Restriction.py:249:9: A500 prefer assertLessEqual() for '<=' comparisons
...
```

So assuming ``test_Restriction.py`` is all fine via #2833, that leaves a few cases in ``test_Pathway.py``, ``test_PDB.py``, and ``common_BioSQL.py`` remaining.

**Comment by trkohler** [comment]
cool! It means I still have some easy things to do! 👯‍♀️

**Comment by trkohler** [comment]
Can I run somehow CI/CD checks in draft PR once again? I added commit to fix errors, but can't test locally - my machine is too weak, it lacks memory and tests run too long.

**Comment by peterjc** [comment]
Since you are only editing the tests, you only need to run that one test - e.g. ``python run_tests.py test_Pathway.py`` which should be quick.

Or I guess you must change the GitHub pull request from draft to normal? The draft thing is quite new so we don't yet have that much experience with it.

**Comment by peterjc** [comment]
Let's generalise this issue to any A500 problem found with ``flake8-assertive``, so equality, less than or equal, etc.

(There is just one case left for ``isinstance``which #2873 should resolve, so we can close #2700)

**Comment by peterjc** [comment]
Updated report, using ``pip install flake8-assertive`` then:

```
$ flake8 --isolated --select A500 .
./test_PDBList.py:41:9: A500 prefer assertGreater() for '>' comparisons
./test_PDBList.py:50:9: A500 prefer assertGreater() for '>' comparisons
./test_NCBI_qblast.py:293:9: A500 prefer assertLessEqual() for '<=' comparisons
./test_NCBI_qblast.py:294:9: A500 prefer assertLessEqual() for '<=' comparisons
./test_NCBI_qblast.py:300:13: A500 prefer assertGreater() for '>' comparisons
./test_NCBI_qblast.py:320:13: A500 prefer assertGreater() for '>' comparisons
./test_Emboss.py:984:13: A500 prefer assertGreater() for '>' comparisons
./test_PopGen_GenePop_EasyController.py:78:9: A500 prefer assertLess() for '<' comparisons
./test_PopGen_GenePop_EasyController.py:112:9: A500 prefer assertLess() for '<' comparisons
./test_PopGen_GenePop_EasyController.py:118:9: A500 prefer assertLess() for '<' comparisons
./test_PAML_tools.py:89:9: A500 prefer assertGreater() for '>' comparisons
./test_PAML_tools.py:110:9: A500 prefer assertGreater() for '>' comparisons
./test_Clustalw_tool.py:98:9: A500 prefer assertEqual() for '==' comparisons
./test_Clustalw_tool.py:115:9: A500 prefer assertEqual() for '==' comparisons
./test_Clustalw_tool.py:122:9: A500 prefer assertEqual() for '==' comparisons
./test_Clustalw_tool.py:160:9: A500 prefer assertEqual() for '==' comparisons
./test_Entrez_online.py:58:9: A500 prefer assertGreater() for '>' comparisons
./test_Entrez_online.py:72:9: A500 prefer assertGreater() for '>' comparisons
./test_Entrez_online.py:226:17: A500 prefer assertGreater() for '>' comparisons
./test_Restriction.py:233:9: A500 prefer assertNotEqual() for '==' comparisons
./test_Restriction.py:235:9: A500 prefer assertEqual() for '!=' comparisons
./test_Restriction.py:242:9: A500 prefer assertGreaterEqual() for '>=' comparisons
./test_Restriction.py:246:9: A500 prefer assertGreater() for '>' comparisons
./test_Restriction.py:249:9: A500 prefer assertLessEqual() for '<=' comparisons
./test_Restriction.py:253:9: A500 prefer assertLess() for '<' comparisons
./test_PDB_vectors.py:144:9: A500 prefer assertLess() for '<' comparisons
./test_PDB_vectors.py:152:9: A500 prefer assertLess() for '<' comparisons
./test_PDB_vectors.py:153:9: A500 prefer assertLess() for '<' comparisons
./test_PDB_vectors.py:162:9: A500 prefer assertLess() for '<' comparisons
./test_Affy.py:39:13: A500 prefer assertGreater() for '>' comparisons
```

**Comment by Logan1x** [comment]
Is this issue still need some work?

**Comment by peterjc** [comment]
Yes please, current master branch gives:

```
$ flake8 --isolated --select A500 .
./test_PDBList.py:41:9: A500 prefer assertGreater() for '>' comparisons
./test_PDBList.py:50:9: A500 prefer assertGreater() for '>' comparisons
./test_NCBI_qblast.py:293:9: A500 prefer assertLessEqual() for '<=' comparisons
./test_NCBI_qblast.py:294:9: A500 prefer assertLessEqual() for '<=' comparisons
./test_NCBI_qblast.py:300:13: A500 prefer assertGreater() for '>' comparisons
./test_NCBI_qblast.py:320:13: A500 prefer assertGreater() for '>' comparisons
./test_Emboss.py:934:13: A500 prefer assertGreater() for '>' comparisons
./test_PopGen_GenePop_EasyController.py:78:9: A500 prefer assertLess() for '<' comparisons
./test_PopGen_GenePop_EasyController.py:112:9: A500 prefer assertLess() for '<' comparisons
./test_PopGen_GenePop_EasyController.py:118:9: A500 prefer assertLess() for '<' comparisons
./test_PAML_tools.py:89:9: A500 prefer assertGreater() for '>' comparisons
./test_PAML_tools.py:110:9: A500 prefer assertGreater() for '>' comparisons
./test_Clustalw_tool.py:98:9: A500 prefer assertEqual() for '==' comparisons
./test_Clustalw_tool.py:115:9: A500 prefer assertEqual() for '==' comparisons
./test_Clustalw_tool.py:122:9: A500 prefer assertEqual() for '==' comparisons
./test_Clustalw_tool.py:160:9: A500 prefer assertEqual() for '==' comparisons
./test_Entrez_online.py:58:9: A500 prefer assertGreater() for '>' comparisons
./test_Entrez_online.py:72:9: A500 prefer assertGreater() for '>' comparisons
./test_Entrez_online.py:226:17: A500 prefer assertGreater() for '>' comparisons
./test_Restriction.py:232:9: A500 prefer assertNotEqual() for '==' comparisons
./test_Restriction.py:234:9: A500 prefer assertEqual() for '!=' comparisons
./test_Restriction.py:241:9: A500 prefer assertGreaterEqual() for '>=' comparisons
./test_Restriction.py:245:9: A500 prefer assertGreater() for '>' comparisons
./test_Restriction.py:248:9: A500 prefer assertLessEqual() for '<=' comparisons
./test_Restriction.py:252:9: A500 prefer assertLess() for '<' comparisons
./test_PDB_internal_coords.py:149:9: A500 prefer assertGreater() for '>' comparisons
./test_PDB_internal_coords.py:150:9: A500 prefer assertGreater() for '>' comparisons
./test_PDB_internal_coords.py:151:9: A500 prefer assertGreater() for '>' comparisons
./test_PDB_vectors.py:147:9: A500 prefer assertLess() for '<' comparisons
./test_PDB_vectors.py:155:9: A500 prefer assertLess() for '<' comparisons
./test_PDB_vectors.py:156:9: A500 prefer assertLess() for '<' comparisons
./test_PDB_vectors.py:165:9: A500 prefer assertLess() for '<' comparisons
./test_Affy.py:39:13: A500 prefer assertGreater() for '>' comparisons
```

Don't do ``test_Restriction.py`` though, see notes above.

**Comment by svalqui** [comment]
I'll work on test_Ace.py (A502).

**Comment by Logan1x** [comment]
@peterjc I couldn't understand anything from your comment. I am really new with this and trying to contribute. Can you ELI5(Explain like I am 5) for this issue?

**Comment by peterjc** [comment]
@Logan1x  Hello. Let me try. Being positive, this will introduce lots of concepts - but there is a lot of background knowledge being assumed here. First of all, our test code has lines like this for comparing variables to expected values:

```python
    self.assertEqual(some_variable, 42)
```

The line above is good from a style point of view, but this next version (although functionally the same) is discouraged:

```python
    self.assertTrue(some_variable == 42)
```

This issue is about improving these kinds of lines in the Biopython code's test - they work, but could be better from a style point of view.

There are tools in most programming languages which can look at your code and report potential bugs (like using a variable which is never initialised) or style issues. One of the first was called "lint" (an English word for the bits of fluff you find in your pocket, meaning unwanted rubbish), so sometimes these tools are called linters. The one we are using for Python is called flake8, and it has lots of additional plugins.

In this issue we are using flake8 with the flake8-assertive plugin, which you can install as follows:

```
pip install flake8 flake8-assertive
```

And then run like this for a single file doing all the checks with the current configuration:

```
$ flake8 example_filename.py
```

Or, as used earlier tell it to ignore the current configuration and just report the A500 errors which are these unit test comparisons checked by flake8-assertive:

```
$ cd Tests/
$ flake8 --isolated --select A500 .
```

Does that help? Or was it what comes next - how to fix one of these stylistic errors, and create a pull request to submit a contribution via GitHub?

**Comment by Logan1x** [comment]
[test.txt](https://github.com/biopython/biopython/files/4993154/test.txt)
this is what I got after running `flake8 --isolated --select A500 .` command. What should be my next steps? @peterjc

**Comment by svalqui** [comment]
@Logan1x ; You could try editing test_PDBList.py, Line 41
https://github.com/biopython/biopython/blob/337f41e786338eb6055918a32afc00976c5a2766/Tests/test_PDBList.py#L41
and replaced with
self.assertGreater(len(entries), 100000)

similar for line 50.
All the best.

**Comment by peterjc** [comment]
Well done @Logan1x, getting flake8 working with the plugin is a good start. Your output differs from mine two weeks ago in sort order (not important) and being more up to date - some issues got fixed by people at the BCC2020 CoFest last week.

The next step is fixing some of these.  @svalqui has explained one example.

You could make tiny edits like this directly in your webbrowser using GitHub (making a pull request), but ideally you would do this on you own computer using an editor, retesting with flake8 and checking the test still works. Is any of that familiar? The git introduction from a Software Carpentry lesson would be useful background - https://swcarpentry.github.io/git-novice/

**Comment by Logan1x** [comment]
@svalqui please review the changes

**Comment by Logan1x** [comment]
@peterjc I am familiar with git and doing changes locally. I need help understanding what changes we are doing. Like I didn't understand the changes I made in latest pr, I just copied it.
was that also a redesign issue?

**Comment by peterjc** [comment]
In this case, using ``.assertGreater(big_value, small_value)`` is considered better style than ``.assertTrue(big_value > small_value)``, see also:

https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertGreater

The behaviour of the code should be identical. More generally, flake8 and related tools can spot things which can be real bugs in code.

**Comment by svalqui** [comment]
@Logan1x , Sorry a bit short of time at the moment, yes that was good, well done.
When I update test files, after a change I run the test, that gives me an indication if the code is affected if it passes means the code is working.
`~/biopython/Tests$ python3 run_tests.py test_PDBList.py`

These changes, pep8; are mainly changes in the style of the code, best practices, makes the code look more standard.
Cheers,

**Comment by sgalpha01** [comment]
Do this issue still need some work?

**Comment by peterjc** [comment]
Yes, there are still a few files to be fixed - using flake8-assertive v1.2.1

```
$ pip install flake8-assertive
$ flake8 --isolated --select A500 .
./Tests/test_NCBI_qblast.py:293:9: A500 prefer assertLessEqual() for '<=' comparisons
./Tests/test_NCBI_qblast.py:294:9: A500 prefer assertLessEqual() for '<=' comparisons
./Tests/test_NCBI_qblast.py:300:13: A500 prefer assertGreater() for '>' comparisons
./Tests/test_NCBI_qblast.py:320:13: A500 prefer assertGreater() for '>' comparisons
./Tests/test_PopGen_GenePop_EasyController.py:78:9: A500 prefer assertLess() for '<' comparisons
./Tests/test_PopGen_GenePop_EasyController.py:112:9: A500 prefer assertLess() for '<' comparisons
./Tests/test_PopGen_GenePop_EasyController.py:118:9: A500 prefer assertLess() for '<' comparisons
./Tests/test_PAML_tools.py:89:9: A500 prefer assertGreater() for '>' comparisons
./Tests/test_PAML_tools.py:110:9: A500 prefer assertGreater() for '>' comparisons
./Tests/test_Entrez_online.py:58:9: A500 prefer assertGreater() for '>' comparisons
./Tests/test_Entrez_online.py:72:9: A500 prefer assertGreater() for '>' comparisons
./Tests/test_Entrez_online.py:226:17: A500 prefer assertGreater() for '>' comparisons
./Tests/test_Restriction.py:232:9: A500 prefer assertNotEqual() for '==' comparisons
./Tests/test_Restriction.py:234:9: A500 prefer assertEqual() for '!=' comparisons
./Tests/test_Restriction.py:241:9: A500 prefer assertGreaterEqual() for '>=' comparisons
./Tests/test_Restriction.py:245:9: A500 prefer assertGreater() for '>' comparisons
./Tests/test_Restriction.py:248:9: A500 prefer assertLessEqual() for '<=' comparisons
./Tests/test_Restriction.py:252:9: A500 prefer assertLess() for '<' comparisons
./Tests/test_PDB_internal_coords.py:149:9: A500 prefer assertGreater() for '>' comparisons
./Tests/test_PDB_internal_coords.py:150:9: A500 prefer assertGreater() for '>' comparisons
./Tests/test_PDB_internal_coords.py:151:9: A500 prefer assertGreater() for '>' comparisons
./Tests/test_SearchIO_model.py:1232:9: A500 prefer assertEqual() for '==' comparisons
./Tests/test_SearchIO_model.py:1234:13: A500 prefer assertEqual() for '==' comparisons
./Tests/test_SearchIO_model.py:1386:9: A500 prefer assertEqual() for '==' comparisons
./Tests/test_SearchIO_model.py:1394:9: A500 prefer assertEqual() for '==' comparisons
./Tests/test_SearchIO_model.py:1395:9: A500 prefer assertEqual() for '==' comparisons
./Tests/test_SearchIO_model.py:1396:9: A500 prefer assertEqual() for '==' comparisons
./Tests/test_PDB_vectors.py:147:9: A500 prefer assertLess() for '<' comparisons
./Tests/test_PDB_vectors.py:155:9: A500 prefer assertLess() for '<' comparisons
./Tests/test_PDB_vectors.py:156:9: A500 prefer assertLess() for '<' comparisons
./Tests/test_PDB_vectors.py:165:9: A500 prefer assertLess() for '<' comparisons
```

(As noted above, best avoid ``test_Restriction.py`` due to non standard object sorting behaviour)

**Comment by sgalpha01** [comment]
Okay, I will try my best. This is going to be my first open-source issue

**Comment by peterjc** [comment]
Good job @sgalpha01 !

If you want to try ``test_Restriction.py`` as well, you'd be most welcome. However, it may much more care since the comparison methods defined are a little strange -- ``__ne__`` is not the inverse of ``__eq__`` for a start:

https://github.com/biopython/biopython/blob/master/Bio/Restriction/Restriction.py#L378
https://github.com/biopython/biopython/blob/master/Bio/Restriction/Restriction.py#L388

**Comment by sgalpha01** [comment]
`flake8 --isolated --select A500 . >> ../TODO.txt` gives the following text file:

> .\Tests\test_NCBI_qblast.py:293:9: A500 prefer assertLessEqual() for '<=' comparisons
> .\Tests\test_NCBI_qblast.py:294:9: A500 prefer assertLessEqual() for '<=' comparisons
> .\Tests\test_NCBI_qblast.py:300:13: A500 prefer assertGreater() for '>' comparisons
> .\Tests\test_NCBI_qblast.py:320:13: A500 prefer assertGreater() for '>' comparisons
> .\Tests\test_Restriction.py:232:9: A500 prefer assertNotEqual() for '==' comparisons
> .\Tests\test_Restriction.py:234:9: A500 prefer assertEqual() for '!=' comparisons
> .\Tests\test_Restriction.py:241:9: A500 prefer assertGreaterEqual() for '>=' comparisons
> .\Tests\test_Restriction.py:245:9: A500 prefer assertGreater() for '>' comparisons
> .\Tests\test_Restriction.py:248:9: A500 prefer assertLessEqual() for '<=' comparisons
> .\Tests\test_Restriction.py:252:9: A500 prefer assertLess() for '<' comparisons

So, there are two files which needs attention. 
I started with `test_NCBI_qblast.py`, and made the required changes.
But, I am not able to commit because of the following error:

```
$ git commit -m "changed assertion styling for test_NCBI_qblast.py"
Check that executables have shebangs.....................................Passed
Check JSON...........................................(no files to check)Skipped
Debug Statements (Python)................................................Passed
Fix End of Files.........................................................Passed
Mixed line ending........................................................Passed
black....................................................................Failed
- hook id: black
- exit code: 1

would reformat Tests\test_NCBI_qblast.py
Oh no! \U0001f4a5 \U0001f494 \U0001f4a5
1 file would be reformatted.

flake8...................................................................Passed
blacken-docs.............................................................Passed
rstcheck.............................................(no files to check)Skipped
doc8.................................................(no files to check)Skipped
Enforce https://doi.org/ style.......................(no files to check)Skipped
```
I am not sure what's the problem. At, first I thought it was editor specific. So, I tried Vim, but no luck.

**Comment by lacrosse91** [comment]
Hello!
Do this issue still need some work?
I'd like to help!

**Comment by peterjc** [comment]
Just ``test_Restriction.py`` left, which is somewhat unusual so I've taken off the "good first issue" tag:

```
$ flake8 --isolated --select A500 .
./test_Restriction.py:232:9: A500 prefer assertNotEqual() for '==' comparisons
./test_Restriction.py:234:9: A500 prefer assertEqual() for '!=' comparisons
./test_Restriction.py:241:9: A500 prefer assertGreaterEqual() for '>=' comparisons
./test_Restriction.py:245:9: A500 prefer assertGreater() for '>' comparisons
./test_Restriction.py:248:9: A500 prefer assertLessEqual() for '<=' comparisons
./test_Restriction.py:252:9: A500 prefer assertLess() for '<' comparisons
```

If @sgalpha01 isn't working on it, you could give it a try. It might go smoothly - just don't change the condition (e.g. don't assume ``a < b`` is the same as ``b > a``, or ``not(a == b)`` is the same as ``a != b`` with these objects).

**Comment by lacrosse91** [comment]
Thanks for reply.
I'll try it.

**Comment by sgalpha01** [comment]
@lacrosse91 , hey you can work on this if you want.

**Comment by lacrosse91** [comment]
@sgalpha01 , Thank you!
I'm going to work on this.

**Comment by peterjc** [comment]
That's all done now:

```
$ flake8 --isolated --select A500 .
(no issues)
```

Thank you everyone - a big team effort.
